app/src/main/java/tech/pauly/findapet/shared/LocationHelper.java

```diff
@@ -0,0 +1,93 @@
+package tech.pauly.findapet.shared;
+
+import android.Manifest;
+import android.content.Context;
+import android.content.pm.PackageManager;
+import android.location.Address;
+import android.location.Criteria;
+import android.location.Geocoder;
+import android.location.Location;
+import android.location.LocationListener;
+import android.location.LocationManager;
+import android.os.Build;
+import android.os.Bundle;
+import android.support.v4.app.ActivityCompat;
+import android.widget.TextView;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Locale;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+import io.reactivex.Observable;
+import io.reactivex.subjects.PublishSubject;
+import tech.pauly.findapet.dependencyinjection.ForApplication;
+
+@Singleton
+public class LocationHelper {
+
+    private Context context;
+    private PublishSubject<String> locationSubject = PublishSubject.create();
+
+    @Inject
+    public LocationHelper(@ForApplication Context context) {
+        this.context = context;
+    }
+
+    public Observable<String> getCurrentLocation() {
+        if (ActivityCompat.checkSelfPermission(context, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
```

Comment:
I'm keeping the check permission logic in the VM for now since it's pretty view-focused and I didn't want to overload this observable stream to do all three permission/location/search. Might change in the future though because it's one of those things that's a little "gotcha" when you want to implement this API on a new screen. Will probably take another look when on the shelters screen.